packages/test/src/test-worker-heartbeats.ts

```diff
@@ -0,0 +1,105 @@
+import test from 'ava';
+import { Subject, firstValueFrom } from 'rxjs';
+import { v4 as uuid4 } from 'uuid';
+import { Context } from '@temporalio/activity';
+import { isolateFreeWorker, Worker } from './mock-native-worker';
+
+async function runActivity(worker: Worker, callback?: () => void) {
+  const taskToken = Buffer.from(uuid4());
+  const p = worker.run();
+  await worker.native.runActivityTask({ taskToken, start: { activityType: 'rapidHeartbeater' } });
+  callback?.();
+  worker.shutdown();
+  await p;
+}
+
+test('Worker stores last heartbeat if flushing is in progress', async (t) => {
+  const subj = new Subject<void>();
+
+  const worker = isolateFreeWorker({
+    taskQueue: 'unused',
+    activities: {
+      async rapidHeartbeater() {
+        Context.current().heartbeat(1);
+        // These details should be overriden by `3`
+        Context.current().heartbeat(2);
+        // This details should be flushed
+        Context.current().heartbeat(3);
+        // Prevent activity from completing
+        await firstValueFrom(subj);
+      },
+    },
+  });
+
+  const heartbeatsSeen = Array<number>();
+  worker.native.activityHeartbeatCallback = (_tt, details) => {
+    heartbeatsSeen.push(details);
+    if (heartbeatsSeen.length === 2) {
+      subj.next();
+    }
+  };
+  await runActivity(worker);
+  t.deepEqual(heartbeatsSeen, [1, 3]);
+});
+
+test('Worker flushes last heartbeat if activity fails', async (t) => {
+  const worker = isolateFreeWorker({
+    taskQueue: 'unused',
+    activities: {
+      async rapidHeartbeater() {
+        Context.current().heartbeat(1);
+        // These details should be overriden by `3`
+        Context.current().heartbeat(2);
+        // This details should be flushed
+        Context.current().heartbeat(3);
+        // Fail
+        throw new Error();
+      },
+    },
+  });
+
+  const heartbeatsSeen = Array<number>();
+  worker.native.activityHeartbeatCallback = (_tt, details) => {
+    heartbeatsSeen.push(details);
+  };
+  await runActivity(worker);
+  t.deepEqual(heartbeatsSeen, [1, 3]);
+});
+
+test('Worker ignores last heartbeat if activity succeeds', async (t) => {
```

Comment:
Why is the last heartbeat ignored on success? I think even on success the last heartbeat should be flushed. I fear people in the UI or future uses of heartbeat details may expect their heartbeat details present. It also makes sense that simply "we guarantee the last heartbeat called will be sent to the server no matter what".

Core has its own buffer and this is a sync call, so I would think it's harmless to flush to Core. Curious though, does Core also discard unflushed heartbeat on success?

Reply:
Core will only flush pending heartbeats on failure.
Heartbeats don't matter on success, details are not persisted in history.

Reply:
I understand this now. I was somehow under the impression heartbeat details could have some value after an activity was done from a UI or history POV.